pkg/apis/core/validation/validation_declarative.go

```diff
@@ -0,0 +1,94 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog/v2"
+
+	fldtest "k8s.io/apimachinery/pkg/util/validation/field/testing"
+	validationmetrics "k8s.io/component-base/metrics/prometheus/validation"
+)
+
+// gatherDeclarativeValidationMismatches compares imperative and declarative validation errors
+// and returns detailed information about any mismatches found.  Errors are compared via type, field, and origin
+func gatherDeclarativeValidationMismatches(imperativeErrs, declarativeErrs field.ErrorList) []string {
+	// Only compare imperative errors that are actually CoveredByDeclarative.
+	coveredImperative := make(field.ErrorList, 0, len(imperativeErrs))
+	for _, err := range imperativeErrs {
+		if err.CoveredByDeclarative {
+			coveredImperative = append(coveredImperative, err)
+		}
+	}
+
+	matcher := fldtest.ErrorMatcher{}.ByType().ByField().ByOrigin().RequireOriginWhenInvalid()
+
+	matchedDeclarative := make([]bool, len(declarativeErrs))
+	var mismatchDetails []string
+
+	// Match each "covered" imperative error to a declarative error.
+	for _, iErr := range coveredImperative {
+		foundMatch := false
+
+		for j, dErr := range declarativeErrs {
+			if !matchedDeclarative[j] && matcher.Matches(iErr, dErr) {
+				matchedDeclarative[j] = true
+				foundMatch = true
+				break
+			}
+		}
+
+		if !foundMatch {
+			mismatchDetails = append(mismatchDetails,
+				fmt.Sprintf(
+					"Imperative validation error marked as CoveredByDeclarative but missing from declarative validation: %s",
+					matcher.Render(iErr),
+				),
+			)
+		}
+	}
+
+	// Any remaining unmatched declarative errors are considered "extra".
+	for j, dErr := range declarativeErrs {
+		if !matchedDeclarative[j] {
+			mismatchDetails = append(mismatchDetails,
+				fmt.Sprintf(
+					"Extra error in declarative validation not found among imperative errors marked as CoveredByDeclarative: %s",
+					matcher.Render(dErr),
+				),
+			)
+		}
+	}
+	return mismatchDetails
+}
+
+// CompareDeclarativeErrorsAndEmitMismatches checks for mismatches between imperative and declarative validation
+// and logs + emits metrics when inconsistencies are found
+func CompareDeclarativeErrorsAndEmitMismatches(imperativeErrs, declarativeErrs field.ErrorList) {
+	mismatchDetails := gatherDeclarativeValidationMismatches(imperativeErrs, declarativeErrs)
+	// Only proceed if there's a mismatch
+	if len(mismatchDetails) > 0 {
+		// Log information about the mismatch
+		for _, detail := range mismatchDetails {
+			klog.Warning(detail)
+		}
+		// Emit the metric
+		validationmetrics.Metrics.EmitDeclarativeValidationMismatchMetric()
```

Comment:
Sorry if this is written down elsewhere, but is this metric documented as counting the # of requests that have a mismatch, or the total number of mismatches encountered?

Reply:
It should be the total # of mismatches encountered:
```
  # declarative_validation_mismatch_total tracks the number of mismatched valdation results and errors
  # when comparing the outputs of hand-written validations vs declarative validations
```

this code was incorrect give the above (it was # of requests).  I have fixed this now